Source/Diagnostics/ReducedDiags/FieldProbe.cpp

```diff
@@ -0,0 +1,250 @@
+                // including ghost cells in the calculation)
+                const auto &arrEx = Ex[mfi].array();
+                const auto &arrEy = Ey[mfi].array();
+                const auto &arrEz = Ez[mfi].array();
+                const auto &arrBx = Bx[mfi].array();
+                const auto &arrBy = By[mfi].array();
+                const auto &arrBz = Bz[mfi].array();
+
+                amrex::ParticleReal Ex_interp = 0._rt, Ey_interp = 0._rt, Ez_interp = 0._rt;
+                amrex::ParticleReal Bx_interp = 0._rt, By_interp = 0._rt, Bz_interp = 0._rt;
+
+                amrex::Array<amrex::Real,3> v_galilean = {{0}};
+                const auto& xyzmin = WarpX::GetInstance().LowerCornerWithGalilean(box,v_galilean,
+                                                                                  lev);
+                const std::array<Real,3>& dx = WarpX::CellSize(lev);
+                amrex::GpuArray<amrex::Real, 3> dx_arr = {dx[0], dx[1], dx[2]};
+                amrex::GpuArray<amrex::Real, 3> xyzmin_arr = {xyzmin[0], xyzmin[1], xyzmin[2]};
+
+                //Interpolating to the probe point
+                doGatherShapeN (x_probe, y_probe, z_probe, Ex_interp, Ey_interp, Ez_interp,
+                                Bx_interp, By_interp, Bz_interp, arrEx, arrEy, arrEz, arrBx,
+                                arrBy, arrBz, Extype, Eytype, Eztype, Bxtype, Bytype, Bztype,
+                                dx_arr, xyzmin_arr, amrex::lbound(box),
+                                WarpX::n_rz_azimuthal_modes, WarpX::nox, false);
+
+                // Either save the interpolated fields or the raw fields depending on the raw_fields flag
+                hv_Ex = raw_fields ? arrEx(i_probe, j_probe, k_probe) : Ex_interp;
+                hv_Ey = raw_fields ? arrEy(i_probe, j_probe, k_probe) : Ey_interp;
+                hv_Ez = raw_fields ? arrEz(i_probe, j_probe, k_probe) : Ez_interp;
+                hv_E = Ex_interp * Ex_interp + Ey_interp * Ey_interp + Ez_interp * Ez_interp;
+                hv_Bx = raw_fields ? arrBx(i_probe, j_probe, k_probe) : Bx_interp;
+                hv_By = raw_fields ? arrBy(i_probe, j_probe, k_probe) : By_interp;
+                hv_Bz = raw_fields ? arrBz(i_probe, j_probe, k_probe) : Bz_interp;
+                hv_B = Bx_interp * Bx_interp + By_interp * By_interp + Bz_interp * Bz_interp;
+
+                probe_proc = amrex::ParallelDescriptor::MyProc();
+            }
+        }
+
+        //All the processors have probe_proc = -1 except the one that contains the point, which
+        //has probe_proc equal to a number >=0. Therefore, ReduceIntMax communicates to all the
+        //processors the rank of the processor which contains the point
+        amrex::ParallelDescriptor::ReduceIntMax(probe_proc);
+
+        if(amrex::ParallelDescriptor::MyProc()==probe_proc){
+            amrex::ParallelDescriptor::Send(&hv_Ex, 1,
```

Comment:
I can imagine two scenarios where the MPI communications could deadlock:

1. The position `(x_probe, y_probe, z_probe)` is not in the physical domain. In this case, I'd expect `probe_proc` to remain -1 and the IO proc will probably wait indefinitely in the recv for a message to arrive. You could add a check that the position is inside the simulation domain before the MFIter loop and MPI communications, and if it is not you could fill the data arrays with zeros and return.
2. The probe_proc is the same as the IO proc. Since you're using blocking MPI calls, this could deadlock in theory (although in practice the exchanged data is small so it'll probably be bufferized). You can possibly skip the communications altogether if the probe_proc is the IO_proc. (using a non blocking send, which I think is Asend in amrex, should work as well)

Reply:
Both are fair points I solved them as follows:

1.  I am now checking if `(x_probe, y_probe, z_probe)` is inside the physical domain. If not I fill the data arrays with zeros.
2. I am skipping the communication if `probe_proc` is the same as the IO_proc